### PR TITLE
[#498] Add option: github-self-hosted while generating the project

### LINKED ISSUE #498: [Feature] Add option: github-self-hosted while generating the project
## Why

The update to the workflow for the iOS template repository must incorporate the newly set up **GitHub self-hosted runner** on a **Mac Mini** from **MacStadium**. The document can be found [HERE](https://www.notion.so/nimblehq/Provision-Custom-MacOS-Runner-for-GitHub-Actions-ace7181d7bf24896804c7e3313c9c956?pvs=4).

The update will add a new option - `github-self-hosted` - in generating project commands in make.sh file
 
## Who Benefits?

- Development Team: Faster build times and improved performance lead to increased productivity and reduced waiting times.
- Project Managers and Stakeholders: Increased reliability and stability facilitate smoother project management and timely delivery.
- Clients and End Users: The updated workflow delivers more robust and bug-free applications, improving the user experience.
- Company: Utilizing the self-hosted runner lowers costs and maximizes credits, reducing dependency on GitHub-hosted runners and providing scalability and flexibility for resource allocation.

### PR DESCRIPTION
**Pull request by ducbm051291**
- Close #498 

## What happened 👀

- Update Script: `SetUpCICDService` to ask for runner type
- We create two folders for each runner type with the same workflows but different runs-on. When the developer chooses a runner type, the chosen type will be copied to the workflow directory then remove the source folder later

## Insight 📝

- N/A

## Proof Of Work 📹

CleanShot 2023-11-16 at 13 56 09

CleanShot 2023-11-16 at 13 55 12
CleanShot 2023-11-16 at 13 55 10
CleanShot 2023-11-16 at 13 55 05
CleanShot 2023-11-16 at 13 54 53

### CODE REVIEW

File: Scripts/Swift/iOSTemplateMaker/Sources/iOSTemplateMaker/SetUpCICDService.swift [hunk 1]
```diff
@@ -33,13 +34,32 @@ struct SetUpCICDService {
 
         switch service {
         case .github:
+            var runnerType: String?
```
Comment (markgravity) [inline Scripts/Swift/iOSTemplateMaker/Sources/iOSTemplateMaker/SetUpCICDService.swift:37]: Can we create an enum for it?

```swift
enum GithubRunnerType: String {
  case macOS = "macos-last"
  case selfHosted = "self-hosted"

  init?(shortName: String) { ... }
}

**Review by markgravity — COMMENTED** [review]
The rest LGTM

File: Scripts/Swift/iOSTemplateMaker/Sources/iOSTemplateMaker/SetUpCICDService.swift [hunk 2]
```diff
@@ -33,13 +34,32 @@ struct SetUpCICDService {
...
+                }
+            }
+            // Use the selected runnerType to configure the workflow
+            if runnerType == "s" {
+                print("Configuring workflow for self-hosted runner")
+                runnerName = "[self-hosted, macOS]"
+            } else {
+                print("Configuring workflow for macOS-latest")
+                runnerName = "macOS-latest"
+            }
```
Comment (blyscuit) [inline Scripts/Swift/iOSTemplateMaker/Sources/iOSTemplateMaker/SetUpCICDService.swift:54]: If we create enum then we can match with enum instead like how we match `CICDService`.

**Comment by blyscuit** [comment]
I am not sure if we will use the same workflows for self-hosted/gh-hosted. Might make sense for `buildAndTest` but `deploy` workflows might have differences between the environment, maybe we add new files for this PR? What do you think?

**Comment by ducbm051291** [comment]
> I am not sure if we will use the same workflows for self-hosted/gh-hosted. Might make sense for `buildAndTest` but `deploy` workflows might have differences between the environment, maybe we add new files for this PR? What do you think?

Good idea 👍

**Review by markgravity — APPROVED** [review]
Perfect to me now ❤️

File: Scripts/Swift/iOSTemplateMaker/Sources/iOSTemplateMaker/SetUpCICDService.swift [hunk 3]
```diff
@@ -33,13 +51,28 @@ struct SetUpCICDService {
...
+            case .selfHosted:
+                fileManager.moveFiles(in: ".github/self_hosted_project_workflows", to: ".github/workflows")
+                fileManager.removeItems(in: ".github/project_workflows")
+                fileManager.removeItems(in: ".github/self_hosted_project_workflows")
+            case .later, .none:
+                print("You can manually setup the runner later.")
+            }
         case .bitrise:
             print("Setting template for Bitrise")
             fileManager.removeItems(in: "codemagic.yaml")
```
[the remark this answers is not in the document]
Comment (ducbm051291) [reply Scripts/Swift/iOSTemplateMaker/Sources/iOSTemplateMaker/SetUpCICDService.swift:78]: Updated 2c46eb2

**Review by blyscuit — APPROVED** [review]

**Pull request merged**